### PR TITLE
Parse root level version instead of relying on token gameversion.

### PR DESCRIPTION
**Pull request by Slendy**
Currently, on Lighthouse the version of a level is determined by the game id in your login ticket but this implementation has some flaws. For example, if you publish an LBP1 level on LBP2 without modifying the level it still publishes as if it were an LBP2 level. Most of the time this shouldn't be an issue but this new method should hopefully be more accurate in determining the game version a level was made in.

Normally I would have put the `CleanupSlotVersionMismatch` logic into an IMigration class because it makes more sense for it to only ever be run once automatically at startup but I am not 100% confident yet to run it. This PR will accurately detect levels from most of the mainline games but may accidentally detect levels from certain vita builds as an lbp2 level or incorrectly categorize beta and alpha builds. For now, I'd rather just have the publishing logic go live so we can test on a smaller scale rather than trying to reversion every slot.

### CODE REVIEW

File: ProjectLighthouse.Servers.GameServer/Controllers/Slots/PublishController.cs [hunk 1]
```diff
@@ -46,14 +45,22 @@ public async Task<IActionResult> StartPublish()
 
         if (string.IsNullOrEmpty(slot.ResourceCollection)) slot.ResourceCollection = slot.RootLevel;
 
+        LbpFile? rootLevel = LbpFile.FromHash(slot.RootLevel);
+
+        if (rootLevel == null) return this.BadRequest();
+
+        GameVersion slotVersion = FileHelper.ParseLevelVersion(rootLevel);
+
+        if (slotVersion == GameVersion.Unknown) slotVersion = gameToken.GameVersion;
```
Comment (jvyden) [inline ProjectLighthouse.Servers.GameServer/Controllers/Slots/PublishController.cs:54]: nit: can prolly be spaced better
```suggestion
        LbpFile? rootLevel = LbpFile.FromHash(slot.RootLevel);
        if (rootLevel == null) return this.BadRequest();

        GameVersion slotVersion = FileHelper.ParseLevelVersion(rootLevel);
        if (slotVersion == GameVersion.Unknown) slotVersion = gameToken.GameVersion;
```

File: ProjectLighthouse.Servers.GameServer/Controllers/Slots/PublishController.cs [hunk 2]
```diff
@@ -101,6 +108,12 @@ public async Task<IActionResult> Publish()
 
         if (rootLevel.FileType != LbpFileType.Level) return this.BadRequest();
 
+        GameVersion slotVersion = FileHelper.ParseLevelVersion(rootLevel);
+
+        slot.GameVersion = slotVersion;
+
+        if (slotVersion == GameVersion.Unknown) slot.GameVersion = gameToken.GameVersion;
+
```
Comment (jvyden) [inline ProjectLighthouse.Servers.GameServer/Controllers/Slots/PublishController.cs:116]: same here
```suggestion
        GameVersion slotVersion = FileHelper.ParseLevelVersion(rootLevel);

        slot.GameVersion = slotVersion;
        if (slotVersion == GameVersion.Unknown) slot.GameVersion = gameToken.GameVersion;

```

File: ProjectLighthouse/Administration/Maintenance/MaintenanceJobs/CleanupSlotVersionMismatchMaintenanceJob.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            if (slotVersion != GameVersion.Unknown)
+            {
+                slot.GameVersion = slotVersion;
+            }
+            
+        }
+
+        await this._database.SaveChangesAsync();
+    }
+}
```
Comment (jvyden) [inline ProjectLighthouse/Administration/Maintenance/MaintenanceJobs/CleanupSlotVersionMismatchMaintenanceJob.cs:34]: For migrations that are run-once you can create a `IMigrationTask`, like [CleanupXMLInjectionMigration](https://github.com/LBPUnion/ProjectLighthouse/blob/main/ProjectLighthouse/Administration/Maintenance/MigrationTasks/CleanupXMLInjectionMigration.cs)

File: ProjectLighthouse/Files/FileHelper.cs [hunk 4]
```diff
@@ -50,6 +51,50 @@ public static bool IsFileSafe(LbpFile file)
...
+        if (file.FileType != LbpFileType.Level)
+        {
+            return GameVersion.Unknown;
+        }
+
+        // Revision numbers borrowed from https://github.com/ennuo/toolkit/blob/main/src/main/java/ennuo/craftworld/resources/structs/Revision.java
+        // LBP3 latest revision is 0x21803F9
+        // LBP2 latest revision is 0x3F8
+        // LBP1 latest revision is 0x272
+        // LBP Vita latest is 0x3E2
```
Comment (jvyden) [inline ProjectLighthouse/Files/FileHelper.cs:65]: Might be worth making these constants?

So for example...
```csharp
const ushort lbpVitaLatest = 0x3E2;
```

File: ProjectLighthouse/Files/FileHelper.cs [hunk 5]
```diff
@@ -50,6 +51,50 @@ public static bool IsFileSafe(LbpFile file)
...
+        GameVersion version = GameVersion.Unknown;
+        if (revision <= 0x272 || revision == 0x2C3)
+        {
+            version = GameVersion.LittleBigPlanet1;
+        }
+        else if (revision == 0x3E2 && branchId == 0x4431 || revision == 0x38B)
+        {
+            version = GameVersion.LittleBigPlanetVita;
+        } else if (revision >> 0x10 != 0)
+        {
```
Comment (jvyden) [inline ProjectLighthouse/Files/FileHelper.cs:87]: nit: formatting
```suggestion
        }
        else if (revision >> 0x10 != 0)
        {
```

File: ProjectLighthouse/Files/FileHelper.cs [hunk 6]
```diff
@@ -50,6 +51,50 @@ public static bool IsFileSafe(LbpFile file)
...
+        uint revision = 0;
+
+        // construct a 32 bit number from 4 individual bytes
+        for (int i = 4; i <= 7; i++)
+        {
+            revision <<= 8;
+            revision |= file.Data[i];
+        }
+        // construct a 16 bit number from 2 individual bytes
+        ushort branchId = (ushort) (file.Data[0xC] << 8 | file.Data[0xD]);
```
Comment (jvyden) [inline ProjectLighthouse/Files/FileHelper.cs:76]: Decimal numbers are a little more clear here imo
```suggestion
        ushort branchId = (ushort) (file.Data[12] << 8 | file.Data[13]);
```

File: ProjectLighthouse/Files/FileHelper.cs [hunk 7]
```diff
@@ -50,6 +51,50 @@ public static bool IsFileSafe(LbpFile file)
         };
     }
 
+    public static GameVersion ParseLevelVersion(LbpFile file)
+    {
+        if (file.FileType != LbpFileType.Level)
+        {
+            return GameVersion.Unknown;
+        }
+
```
Comment (jvyden) [inline ProjectLighthouse/Files/FileHelper.cs:60]:
```suggestion
        if (file.FileType != LbpFileType.Level) return GameVersion.Unknown;
```

**Review by jvyden — CHANGES_REQUESTED** [review]
Looks good! Just a couple concerns and the usual nitpicks...

**Comment by jvyden** [comment]
Also, this would be good to unit test

**Comment by jvyden** [comment]
@ennuo This looks fine to me, but you're the expert when it comes to things like this. Do you mind reviewing this PR just incase there are any concerns with this implementation?

**Comment by jvyden** [comment]
Also, @AShifter would you mind testing this?

**Comment by ennuo** [comment]
Just a few notes

Branch descriptions weren't added until 0x271, so you're potentially reading random data there for earlier revisions, it probably wouldn't cause a problem since you're only checking the branch for Vita, but something to be aware of.

I don't think anyone would be funny enough to upload text-based levels, LVLt, but might be something good to check for since you're reading binary data here, as well as checking the length of the data in case someone's uploading random garbage that just happens to start with LVL.

Not sure where 0x38b comes from for Vita revisions, should always be 0x3e2, unless that's some earlier build before they branched that I'm not aware of?

If you're including Deploy build levels as part of LBP1, there's quite a few files between 0x272 and 0x2c3 in the files themselves that someone might upload, somewhat of a disconnect if you consider that LBP2, but user created ones LBP1.

**Comment by Slendy** [comment]
I was downloading random levels that had been uploaded to Lighthouse to test this and one of the vita ones had 38b with no branch descriptor, so I'm assuming it was some sort of earlier build.

**Comment by jvyden** [comment]
Kinda related: do you need a full dump of all the levels on Lighthouse? That's something Aidan and I were working on a few days ago.

**Comment by jvyden** [comment]
> so I'm assuming it was some sort of earlier build.
Also, @turecross321 might be able to explain this in more detail? He's familiar with a lot of the quirks of LBP Vita, though I wouldn't expect him to respond soon cause he's on vacation lol

**Review by jvyden — CHANGES_REQUESTED** [review]
Looking better! I think it's definitely worth handling the edge-cases Aidan mentioned, and then we can test & merge. More on testing later...

Anyways, here's what we should handle:

- [x] Only check branch descriptors if revision is >= 0x271
- [x] Block text format files entirely? Depends on if they're actually used often.

Not entirely sure if Deploy build is worth handling. Do deploy levels have special branch descriptors? I might be misunderstanding the problem

**Comment by jvyden** [comment]
Anyways, as for testing here's a general plan:

- Clone Beacon (production lighthouse) database
- Run the migration job on the cloned database
- Compare `GameVersion`s on the clone to the ones on the production database
- Manually review each difference

If everything checks out we should be good to merge. Worst case scenario we find some bugs.

**Comment by ennuo** [comment]
You probably will never actually encounter LVLt, it's just a byproduct of the fact that every resource in the game has a text variant.

As for deploy, I mentioned it because the check for LBP1 above is <=0x272 or 0x2c3, 0x2c3 being levels published from the Deploy build, although maybe it would probably be best to consider them LBP2 or just disallow them, since they won't work in LBP1 and the fact that it's not a release build

**Comment by Slendy** [comment]
If you wanted to handle the text variants you would probably want to do it in DetermineFileType and just check the 4th byte.

**Comment by jvyden** [comment]
Did some digging. There is only one text variant uploaded to lighthouse, with the type of `SSP`. Not entirely sure what this is:

[a060944abfa19e3c1ec0a3794b7078a51242c6e9.txt](https://github.com/LBPUnion/ProjectLighthouse/files/8983567/a060944abfa19e3c1ec0a3794b7078a51242c6e9.txt)

Theres also a `GTFs` file, and a `GTFS` file. Not sure what these are either. Heres what `file` has to say about all of them:

```
[jvyden@the-overcooler r]$ file a060944abfa19e3c1ec0a3794b7078a51242c6e9
a060944abfa19e3c1ec0a3794b7078a51242c6e9: ASCII text
[jvyden@the-overcooler r]$ file 41eb86cee55af8485e2d8db79631358aff3f2675
41eb86cee55af8485e2d8db79631358aff3f2675: data
[jvyden@the-overcooler r]$ file 552381cc2c21789c7094b18a54d33122981dcc29
552381cc2c21789c7094b18a54d33122981dcc29: GLS_BINARY_LSB_FIRST
```

So with all that in mind I think we should be fine to just block text variant files.

And for reference, here's a quick script I wrote in NodeJS to check:

```js
#!/usr/bin/node

const dir = '../ProjectLighthouse/r/';

const fs = require('fs');
const files = fs.readdirSync(dir);

const offset = 3;

async function run() {
    for(let file of files) {
        const chunks = [];
        for await(let chunk of fs.createReadStream(dir + file, {start: 0, end: offset})) {
            chunks.push(chunk);
        }

        const fileData = Buffer.concat(chunks);

        if(fileData[offset] <= 32 || fileData[offset] >= 126) continue; // get bytes in binary range
        const type = String.fromCharCode(fileData[offset]);

        if(type == 'b' || type == 'G') continue; // filter out typical binary files and PN(G)s

        console.log(file, fileData[offset], type);
    }
}

run()
```

**Comment by ennuo** [comment]
> There is only one text variant uploaded to lighthouse, with the type of `SSP`. Not entirely sure what this is:

Softbody physics settings, somebody probably just uploaded a model with hash dependency references.

> Theres also a `GTFs` file, and a `GTFS` file. Not sure what these are either.

They're just Vita texture types to distinguish from PS3 GTF files, `s` probably just to indicate the data is swizzled, `S` has additional header metadata. I don't think Vita usually generates these files for UGC content, so probably modded content as well.

**Comment by turecross321** [comment]
> > so I'm assuming it was some sort of earlier build.
> 
> Also, @turecross321 might be able to explain this in more detail? He's familiar with a lot of the quirks of LBP Vita, though I wouldn't expect him to respond soon cause he's on vacation lol

Might be the Vita part of a Cross Controller level? Not sure.

**Comment by TorutheRedFox** [comment]
SSPt is a physics definition for softbodies (has .sph extension in game data)

it defines stuff like general rigidity and wind info such as direction, phase, frequency and strength

**Comment by jvyden** [comment]
Okay, I just ran the numbers. Here's what this changed on production:

<details>
<summary>Spoiler to prevent spam</summary>
<br>

```
Slot 33 changed: 1 -> 0
Slot 35 changed: 1 -> 0
Slot 37 changed: 1 -> 0
Slot 40 changed: 1 -> 0
Slot 2940 changed: 1 -> 0
Slot 2965 changed: 1 -> 0
Slot 2978 changed: 1 -> 0
Slot 2995 changed: 1 -> 2
Slot 2996 changed: 1 -> 2
Slot 2997 changed: 1 -> 2
Slot 2998 changed: 1 -> 2
Slot 3009 changed: 1 -> 0
Slot 3036 changed: 1 -> 0
Slot 3048 changed: 1 -> 0
Slot 3060 changed: 1 -> 0
Slot 3091 changed: 1 -> 0
Slot 3093 changed: 1 -> 0
Slot 3102 changed: 1 -> 0
Slot 3188 changed: 0 -> 1
Slot 3249 changed: 1 -> 0
Slot 3383 changed: 0 -> 1
Slot 3403 changed: 1 -> 0
Slot 3405 changed: 1 -> 0
Slot 3408 changed: 1 -> 0
Slot 3410 changed: 1 -> 0
Slot 3465 changed: 1 -> 0
Slot 3470 changed: 1 -> 0
Slot 3490 changed: 1 -> 0
Slot 3515 changed: 1 -> 3
Slot 3518 changed: 1 -> 3
Slot 3538 changed: 0 -> 1
Slot 3586 changed: 0 -> 1
Slot 3589 changed: 1 -> 0
Slot 3598 changed: 1 -> 3
Slot 3613 changed: 0 -> 1
Slot 3614 changed: 0 -> 1
Slot 3626 changed: 1 -> 0
Slot 3627 changed: 1 -> 0
Slot 3652 changed: 1 -> 0
Slot 3666 changed: 1 -> 0
Slot 3667 changed: 1 -> 0
Slot 3675 changed: 1 -> 0
Slot 3676 changed: 1 -> 0
Slot 3677 changed: 1 -> 0
Slot 3678 changed: 1 -> 0
Slot 3679 changed: 1 -> 0
Slot 3680 changed: 1 -> 0
Slot 3681 changed: 1 -> 0
Slot 3682 changed: 1 -> 0
Slot 3683 changed: 1 -> 0
Slot 3684 changed: 1 -> 0
Slot 3685 changed: 1 -> 0
Slot 3686 changed: 1 -> 0
Slot 3699 changed: 1 -> 0
Slot 3700 changed: 1 -> 0
Slot 3701 changed: 1 -> 0
Slot 3703 changed: 1 -> 0
Slot 3706 changed: 1 -> 0
Slot 3707 changed: 1 -> 0
Slot 3731 changed: 0 -> 1
Slot 3740 changed: 0 -> 1
Slot 3766 changed: 0 -> 1
Slot 3768 changed: 0 -> 1
Slot 3774 changed: 0 -> 1
Slot 3778 changed: 1 -> 0
Slot 3779 changed: 1 -> 0
Slot 3787 changed: 1 -> 0
Slot 3788 changed: 1 -> 0
Slot 3805 changed: 0 -> 1
Slot 3806 changed: 0 -> 1
Slot 3809 changed: 0 -> 1
Slot 3898 changed: 0 -> 1
Slot 3906 changed: 1 -> 0
Slot 3915 changed: 1 -> 0
Slot 3921 changed: 0 -> 1
Slot 3960 changed: 1 -> 0
Slot 3964 changed: 0 -> 1
Slot 3965 changed: 0 -> 1
Slot 3997 changed: 1 -> 0
Slot 3998 changed: 1 -> 0
Slot 4003 changed: 1 -> 0
Slot 4004 changed: 1 -> 0
Slot 4006 changed: 1 -> 0
Slot 4007 changed: 1 -> 0
Slot 4011 changed: 1 -> 0
Slot 4020 changed: 1 -> 0
Slot 4036 changed: 1 -> 0
Slot 4072 changed: 1 -> 3
Slot 4073 changed: 1 -> 3
Slot 4074 changed: 1 -> 3
Slot 4083 changed: 1 -> 3
Slot 4150 changed: 1 -> 0
Slot 4157 changed: 1 -> 0
Slot 4173 changed: 1 -> 0
Slot 4177 changed: 1 -> 3
Slot 4190 changed: 1 -> 3
Slot 4236 changed: 1 -> 3
Slot 4238 changed: 1 -> 0
Slot 4239 changed: 1 -> 0
Slot 4246 changed: 1 -> 0
Slot 4269 changed: 1 -> 3
Slot 4290 changed: 1 -> 3
Slot 4354 changed: 1 -> 0
Slot 4355 changed: 1 -> 0
Slot 4434 changed: 1 -> 0
Slot 4444 changed: 1 -> 0
Slot 4486 changed: 1 -> 0
Slot 4574 changed: 0 -> 1
Slot 4575 changed: 0 -> 1
Slot 4576 changed: 0 -> 1
Slot 4577 changed: 0 -> 1
Slot 4578 changed: 1 -> 0
Slot 4582 changed: 1 -> 0
Slot 4583 changed: 1 -> 0
Slot 4584 changed: 1 -> 0
Slot 4592 changed: 1 -> 0
```

</details>

**Comment by jvyden** [comment]
3188 and beyond are going to have to be checked manually, fairly certain all the ones before are safe; they were uploaded before `GameVersion` existed.

**Comment by jvyden** [comment]
> Not sure where 0x38b comes from for Vita revisions, should always be 0x3e2, unless that's some earlier build before they branched that I'm not aware of?

@ennuo So we (Slendy and I) think we found out what happened here, and I'd like to share what came of our discussion on Discord.

@turecross321 uploaded a Vita level with the LBP2 retail background and it presumably came with some jank. Then, Slendy downloaded this level and looked at it to determine what a Vita level looks like it, and got the wrong idea from the cursed level.

So, when we tested this change it caused a bunch of LBP2 Beta/Pre-Alpha levels to be converted to Vita levels.

Here's the plan: let's try this again using 0x3e2 as Aidan suggested. Hopefully we won't see any problems with Beta levels this time, and we'll be able to merge.

As an additional sidenote, we also discussed Deploy and I think we're okay with sacrificing those levels for now. It's a niche area that's pretty hard to work with considering it shares the same TitleID as LBP1. 

If we find an easy way we'll happily work with Deploy, but it's definitely not a priority.

Anyways, with that out of all the way, all that's left is to fix Beta levels going into Vita and this looks ready to merge!

**Comment by jvyden** [comment]
Here's the new list off the latest commit.

<details>
<summary>Spoiler to prevent spam</summary>
<br>

```
Slot 33 changed: 1 -> 0
Slot 35 changed: 1 -> 0
Slot 37 changed: 1 -> 0
Slot 40 changed: 1 -> 0
Slot 2940 changed: 1 -> 0
Slot 2965 changed: 1 -> 0
Slot 2978 changed: 1 -> 0
Slot 2995 changed: 1 -> 2
Slot 2996 changed: 1 -> 2
Slot 2997 changed: 1 -> 2
Slot 2998 changed: 1 -> 2
Slot 3009 changed: 1 -> 0
Slot 3036 changed: 1 -> 0
Slot 3048 changed: 1 -> 0
Slot 3060 changed: 1 -> 0
Slot 3091 changed: 1 -> 0
Slot 3093 changed: 1 -> 0
Slot 3102 changed: 1 -> 0
Slot 3188 changed: 0 -> 1
Slot 3249 changed: 1 -> 0
Slot 3383 changed: 0 -> 1
Slot 3403 changed: 1 -> 0
Slot 3405 changed: 1 -> 0
Slot 3408 changed: 1 -> 0
Slot 3410 changed: 1 -> 0
Slot 3465 changed: 1 -> 0
Slot 3470 changed: 1 -> 0
Slot 3490 changed: 1 -> 0
Slot 3531 changed: 3 -> 1
Slot 3538 changed: 0 -> 1
Slot 3586 changed: 0 -> 1
Slot 3589 changed: 1 -> 0
Slot 3613 changed: 0 -> 1
Slot 3614 changed: 0 -> 1
Slot 3626 changed: 1 -> 0
Slot 3627 changed: 1 -> 0
Slot 3652 changed: 1 -> 0
Slot 3666 changed: 1 -> 0
Slot 3667 changed: 1 -> 0
Slot 3675 changed: 1 -> 0
Slot 3676 changed: 1 -> 0
Slot 3677 changed: 1 -> 0
Slot 3678 changed: 1 -> 0
Slot 3679 changed: 1 -> 0
Slot 3680 changed: 1 -> 0
Slot 3681 changed: 1 -> 0
Slot 3682 changed: 1 -> 0
Slot 3683 changed: 1 -> 0
Slot 3684 changed: 1 -> 0
Slot 3685 changed: 1 -> 0
Slot 3686 changed: 1 -> 0
Slot 3699 changed: 1 -> 0
Slot 3700 changed: 1 -> 0
Slot 3701 changed: 1 -> 0
Slot 3703 changed: 1 -> 0
Slot 3706 changed: 1 -> 0
Slot 3707 changed: 1 -> 0
Slot 3731 changed: 0 -> 1
Slot 3740 changed: 0 -> 1
Slot 3766 changed: 0 -> 1
Slot 3768 changed: 0 -> 1
Slot 3774 changed: 0 -> 1
Slot 3778 changed: 1 -> 0
Slot 3779 changed: 1 -> 0
Slot 3787 changed: 1 -> 0
Slot 3788 changed: 1 -> 0
Slot 3805 changed: 0 -> 1
Slot 3806 changed: 0 -> 1
Slot 3809 changed: 0 -> 1
Slot 3898 changed: 0 -> 1
Slot 3906 changed: 1 -> 0
Slot 3915 changed: 1 -> 0
Slot 3921 changed: 0 -> 1
Slot 3960 changed: 1 -> 0
Slot 3964 changed: 0 -> 1
Slot 3965 changed: 0 -> 1
Slot 3997 changed: 1 -> 0
Slot 3998 changed: 1 -> 0
Slot 4003 changed: 1 -> 0
Slot 4004 changed: 1 -> 0
Slot 4006 changed: 1 -> 0
Slot 4007 changed: 1 -> 0
Slot 4011 changed: 1 -> 0
Slot 4020 changed: 1 -> 0
Slot 4036 changed: 1 -> 0
Slot 4150 changed: 1 -> 0
Slot 4157 changed: 1 -> 0
Slot 4173 changed: 1 -> 0
Slot 4238 changed: 1 -> 0
Slot 4239 changed: 1 -> 0
Slot 4246 changed: 1 -> 0
Slot 4354 changed: 1 -> 0
Slot 4355 changed: 1 -> 0
Slot 4434 changed: 1 -> 0
Slot 4444 changed: 1 -> 0
Slot 4486 changed: 1 -> 0
Slot 4574 changed: 0 -> 1
Slot 4575 changed: 0 -> 1
Slot 4576 changed: 0 -> 1
Slot 4577 changed: 0 -> 1
Slot 4578 changed: 1 -> 0
Slot 4582 changed: 1 -> 0
Slot 4583 changed: 1 -> 0
Slot 4584 changed: 1 -> 0
Slot 4592 changed: 1 -> 0
Slot 4610 changed: 1 -> 0
```

</summary>

**Comment by Slendy** [comment]
I think most of the ones that are lbp1 -> lbp2 are levels from deploy or similar nonstandard versions and the ones that went from lbp2 to lbp1 are lbp1 levels that were imported and published using lbp2

**Review by jvyden — APPROVED** [review]
Tested a bunch of levels, and we're both ready to pull the trigger. YOLO.

**Comment by jvyden** [comment]
Deployment went smoothly. Thanks to everyone involved!
